KANBAN-858: Add pg_trgm GIN index for fuzzy search performance

christabone opened the pull request:
Implements PostgreSQL pg_trgm extension and GIN index on slotannotation.displaytext to improve fuzzy search performance for the AI curation tool.

## Performance Impact
- **Before**: ~2 seconds per fuzzy search query (sequential scan over ~18M rows)
- **After**: ~20-50ms per fuzzy search query (GIN index scan)
- **Improvement**: 50-100x faster queries

## Changes
1. **Flyway Migration** (v0.43.0.4__add_pg_trgm_fuzzy_search_index.sql)
   - Enables pg_trgm extension
   - Drops existing B-tree index on displaytext
   - Creates GIN index with gin_trgm_ops operator class
   - Uses CREATE INDEX CONCURRENTLY to avoid blocking operations

2. **Entity Update** (SlotAnnotation.java)
   - Removed @Index annotation for displaytext column
   - Index now managed exclusively via Flyway migration

## Trade-offs
- Additional disk space: ~few hundred MB for GIN index
- Slightly slower INSERT/UPDATE operations on displaytext column
- Net benefit: Dramatically improved read performance

## Deployment Considerations
- Migration uses CREATE INDEX CONCURRENTLY to minimize blocking
- Recommended to run during low-traffic window
- Estimated index creation time: 10-30 minutes for ~18M rows
- Requires A-Team approval before deployment

Implements requirements discussed with @olinblodgett for managing index via Java/Hibernate (Flyway).

christabone committed:
KANBAN-858: Add pg_trgm GIN index for fuzzy search performance

Implements PostgreSQL pg_trgm extension and GIN index on slotannotation.displaytext
to improve fuzzy search performance from ~2s to ~20-50ms (50-100x faster).

Changes:
- Add Flyway migration v0.43.0.4 to enable pg_trgm and create GIN index
- Remove @Index annotation from SlotAnnotation.java (managed via migration)
- Use CREATE INDEX CONCURRENTLY to avoid blocking operations

oblodgett commented on src/main/resources/db/migration/v0.43.0.5__add_pg_trgm_fuzzy_search_index.sql:
```diff
@@ -0,0 +1,15 @@
+-- Create GIN index with pg_trgm operator class for fuzzy search performance
+-- This improves fuzzy search queries from ~2s to ~20-50ms (50-100x improvement)
+-- Using CONCURRENTLY to avoid blocking table operations during index creation
+CREATE INDEX CONCURRENTLY slotannotation_displaytext_trgm_idx
```
@christabone the index name in java needs to match the name in the DB

christabone replied:
@oblodgett The Java @Index annotation name now matches the database index name created by the Flyway migration. Both are `slotannotation_displaytext_trgm_idx`. Does that work?

christabone committed:
Clarify displaytext index comment in SlotAnnotation entity

Updated comment to explicitly state the old index was removed
and is now managed via Flyway migration for pg_trgm GIN index.

christabone committed:
Add @Index annotation to match database index name

Per Olin's feedback, the index name in Java needs to match the
name in the DB. Added @Index annotation for slotannotation_displaytext_trgm_idx
to match the GIN index created by Flyway migration.

christabone committed:
Fix Flyway mixed transaction error in migration

Changed DROP INDEX to use CONCURRENTLY to match the non-transactional
mode of CREATE INDEX CONCURRENTLY. This prevents Flyway from detecting
mixed transactional/non-transactional statements in the same migration.

Fixes integration test failure caused by FlywayException.

christabone committed:
Split pg_trgm migration into two files to fix Flyway error

Flyway cannot mix transactional and non-transactional statements in
the same migration file. Split into:

1. v0.43.0.4__enable_pg_trgm_extension.sql (transactional)
   - Creates pg_trgm extension

2. v0.43.0.5__add_pg_trgm_fuzzy_search_index.sql (non-transactional)
   - Drops old B-tree index using CONCURRENTLY
   - Creates new GIN trigram index using CONCURRENTLY

This fixes the integration test failure caused by mixed transaction modes.

christabone committed:
Fix migration: use CONCURRENTLY for DROP INDEX, remove unsupported IF NOT EXISTS

Changes:
- Added CONCURRENTLY to DROP INDEX (required for non-transactional mode)
- Removed IF NOT EXISTS from CREATE INDEX CONCURRENTLY (not supported in PostgreSQL)
- Added public schema prefix to DROP INDEX for clarity

Both statements now use CONCURRENTLY, ensuring compatibility with flyway:noTransaction.

christabone commented:
Running tests locally to figure out the issue here...

christabone committed:
Simplify pg_trgm index migration to use standard CREATE INDEX

Remove CONCURRENTLY keyword and flyway:noTransaction directive to align
with the 900+ existing CREATE INDEX statements in the migration history.

The previous approach using CREATE INDEX CONCURRENTLY with
flyway:noTransaction caused a Flyway self-deadlock during integration
tests, where Flyway's idle metadata connection blocked the concurrent
index creation indefinitely.

This simplified approach:
- Runs in a standard Flyway transaction (no deadlock)
- Matches project conventions (all other indexes use plain CREATE INDEX)
- Tests pass in ~2 minutes instead of timing out at 10+ minutes
- Production deployment will briefly lock the table during index creation
  (estimated 10-15 minutes on 18M rows), consistent with other migrations

Fixes KANBAN-858

christabone commented:
@oblodgett I think this is ready to go now.

cmpich reviewed (approved):
It looks good. I don't have experience with the GIN index but we can try it out on alpha first and see it working.
